// @spectrum-web-components/meter/stories/meter.stories.js
"use strict";
import { html } from "@spectrum-web-components/base";
import "@spectrum-web-components/meter/sp-meter.js";
export default {
  title: "Meter",
  component: "sp-meter"
};
const makeOverBackground = (story) => html`
    <div
        style="background-color: var(--spectrum-global-color-seafoam-600); color: var(--spectrum-global-color-seafoam-600); padding: var(--spectrum-global-dimension-size-175) var(--spectrum-global-dimension-size-250); display: inline-block"
    >
        ${story}
    </div>
`;
export const sideLabel = () => {
  return html`
        <sp-meter side-label progress="50">Storage Space</sp-meter>
    `;
};
export const negative = () => {
  return html`
        <sp-meter variant="negative" progress="95">Storage Space</sp-meter>
    `;
};
export const notice = () => {
  return html`
        <sp-meter variant="notice" progress="73">Storage Space</sp-meter>
    `;
};
export const positive = () => {
  return html`
        <sp-meter variant="positive" progress="50">Storage Space</sp-meter>
    `;
};
export const staticWhite = () => {
  return makeOverBackground(
    html`
            <sp-meter static="white" progress="50">Storage Space</sp-meter>
        `
  );
};
