// @spectrum-web-components/progress-bar/stories/progress-bar.stories.js
"use strict";
import { html } from "@spectrum-web-components/base";
import "@spectrum-web-components/progress-bar/sp-progress-bar.js";
export default {
  title: "Progress Bar",
  component: "sp-progress-bar"
};
export const label = () => {
  return html`
        <sp-progress-bar label="Loading" progress="50"></sp-progress-bar>
    `;
};
export const indeterminate = () => {
  return html`
        <sp-progress-bar label="Loading" indeterminate></sp-progress-bar>
    `;
};
export const sideLabel = () => {
  return html`
        <sp-progress-bar
            side-label
            label="Loading"
            progress="50"
        ></sp-progress-bar>
    `;
};
export const sideIndeterminate = () => {
  return html`
        <sp-progress-bar
            side-label
            label="Loading"
            indeterminate
        ></sp-progress-bar>
    `;
};
const makeOverBackground = (story) => html`
    <div
        style="background-color: var(--spectrum-global-color-seafoam-600); color: var(--spectrum-global-color-seafoam-600); padding: var(--spectrum-global-dimension-size-175) var(--spectrum-global-dimension-size-250); display: inline-block"
    >
        ${story}
    </div>
`;
export const staticWhite = () => {
  return makeOverBackground(html`
        <sp-progress-bar progress="50" static="white"></sp-progress-bar>
    `);
};
export const staticWhiteLabel = () => {
  return makeOverBackground(html`
        <sp-progress-bar
            label="Loading"
            progress="50"
            static="white"
        ></sp-progress-bar>
    `);
};
export const staticWhiteIndeterminate = () => {
  return makeOverBackground(html`
        <sp-progress-bar
            label="Loading"
            indeterminate
            static="white"
        ></sp-progress-bar>
    `);
};
export const staticWhiteSideLabel = () => {
  return makeOverBackground(html`
        <sp-progress-bar
            label="Loading"
            progress="50"
            static="white"
            side-label
        ></sp-progress-bar>
    `);
};
export const staticWhiteSideLabelIndeterminate = () => {
  return makeOverBackground(html`
        <sp-progress-bar
            label="Loading"
            indeterminate
            static="white"
            side-label
        ></sp-progress-bar>
    `);
};
