// @spectrum-web-components/split-button/stories/split-button-primary-field.stories.js
"use strict";
import { ElementSizes } from "@spectrum-web-components/base";
import { args, argTypes, renderSplitButtonSet, splitbutton } from "./index.js";
import { openSplitButtonDecorator } from "./helpers.js";
import "./helpers.js";
import "@spectrum-web-components/split-button/sp-split-button.js";
import "@spectrum-web-components/menu/sp-menu-item.js";
const variant = "primary";
const type = "field";
export default {
  title: "Split Button/Primary/Field",
  component: "sp-split-button",
  args: {
    ...args,
    variant,
    type
  },
  argTypes
};
export const s = (args2) => renderSplitButtonSet(args2);
s.args = { size: ElementSizes.s };
export const m = (args2) => renderSplitButtonSet(args2);
m.args = { size: ElementSizes.m };
export const l = (args2) => renderSplitButtonSet(args2);
l.args = { size: ElementSizes.l };
export const XL = (args2) => renderSplitButtonSet(args2);
XL.args = { size: ElementSizes.xl };
export const open = (args2) => splitbutton(args2);
open.args = { open: true };
open.decorators = [openSplitButtonDecorator];
