// @spectrum-web-components/avatar/src/Avatar.dev.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __decorateClass = (decorators, target, key, kind) => {
  var result = kind > 1 ? void 0 : kind ? __getOwnPropDesc(target, key) : target;
  for (var i = decorators.length - 1, decorator; i >= 0; i--)
    if (decorator = decorators[i])
      result = (kind ? decorator(target, key, result) : decorator(result)) || result;
  if (kind && result)
    __defProp(target, key, result);
  return result;
};
import {
  html,
  SpectrumElement
} from "@spectrum-web-components/base";
import { property } from "@spectrum-web-components/base/src/decorators.js";
import { ifDefined } from "@spectrum-web-components/base/src/directives.js";
import avatarStyles from "./avatar.css.js";
const validSizes = [50, 75, 100, 200, 300, 400, 500, 600, 700];
const defaultSize = validSizes[2];
export class Avatar extends SpectrumElement {
  constructor() {
    super(...arguments);
    this.label = "";
    this.src = "";
    this._size = defaultSize;
  }
  static get styles() {
    return [avatarStyles];
  }
  get size() {
    return this._size;
  }
  set size(value) {
    const size = value;
    const validSize = validSizes.includes(size) ? size : defaultSize;
    if (validSize) {
      this.setAttribute("size", `${validSize}`);
    }
    if (this._size === validSize) {
      return;
    }
    const oldSize = this._size;
    this._size = validSize;
    this.requestUpdate("size", oldSize);
  }
  render() {
    return html`
            <img
                class="image"
                alt=${ifDefined(this.label || void 0)}
                src=${this.src}
            />
        `;
  }
  firstUpdated(changes) {
    super.firstUpdated(changes);
    if (!this.hasAttribute("size")) {
      this.setAttribute("size", `${this.size}`);
    }
  }
}
__decorateClass([
  property()
], Avatar.prototype, "label", 2);
__decorateClass([
  property()
], Avatar.prototype, "src", 2);
__decorateClass([
  property({ type: Number, reflect: true })
], Avatar.prototype, "size", 1);
