// @swc-uxp-internal/menu/stories/submenu.stories.js
"use strict";
import { html, render } from "@spectrum-web-components/base";
import "@spectrum-web-components/action-menu/sp-action-menu.js";
import "@spectrum-web-components/menu/sp-menu-item.js";
import "@spectrum-web-components/menu/sp-menu-divider.js";
import "@spectrum-web-components/menu/sp-menu-group.js";
import { openOverlay, VirtualTrigger } from "@spectrum-web-components/overlay";
import "@spectrum-web-components/popover/sp-popover.js";
import "@spectrum-web-components/icons-workflow/icons/sp-icon-show-menu.js";
export default {
  component: "sp-menu",
  title: "Menu/Submenu"
};
function nextFrame() {
  return new Promise((res) => requestAnimationFrame(() => res()));
}
class SubmenuReady extends HTMLElement {
  constructor() {
    super();
    this.handleMenuOpened = async (event) => {
      await nextFrame();
      await event.target.updateComplete;
      const submenu2 = document.querySelector("#submenu-item-1");
      if (!submenu2) {
        return;
      }
      submenu2.addEventListener("sp-opened", this.handleSubmenuOpened, {
        once: true
      });
      submenu2.dispatchEvent(
        new PointerEvent("pointerenter", { bubbles: true, composed: true })
      );
    };
    this.handleSubmenuOpened = async (event) => {
      await nextFrame();
      await event.target.updateComplete;
      const submenu2 = document.querySelector("#submenu-item-2");
      if (!submenu2) {
        return;
      }
      submenu2.addEventListener("sp-opened", this.handleSubmenuChildOpened, {
        once: true
      });
      submenu2.dispatchEvent(
        new PointerEvent("pointerenter", { bubbles: true, composed: true })
      );
    };
    this.handleSubmenuChildOpened = async (event) => {
      await nextFrame();
      await event.target.updateComplete;
      this.ready(true);
    };
    this.readyPromise = Promise.resolve(false);
    this.readyPromise = new Promise((res) => {
      this.ready = res;
      this.setup();
    });
  }
  async setup() {
    await nextFrame();
    const menu = document.querySelector(`sp-action-menu`);
    menu.addEventListener("sp-opened", this.handleMenuOpened, {
      once: true
    });
    menu.open = true;
  }
  get updateComplete() {
    return this.readyPromise;
  }
}
customElements.define("submenu-ready", SubmenuReady);
const submenuDecorator = (story) => {
  return html`
        ${story()}
        <submenu-ready></submenu-ready>
    `;
};
export const submenu = () => {
  const getValueEls = () => {
    return {
      root: document.querySelector("#root-value"),
      first: document.querySelector("#first-value"),
      second: document.querySelector("#second-value")
    };
  };
  const clearValues = () => {
    const valueEls = getValueEls();
    valueEls.root.textContent = "";
    valueEls.first.textContent = "";
    valueEls.second.textContent = "";
  };
  const handleRootChange = (event) => {
    const valueEls = getValueEls();
    valueEls.root.textContent = event.target.value;
  };
  const handleFirstDescendantChange = (event) => {
    const valueEls = getValueEls();
    valueEls.first.textContent = event.target.selected[0] || "";
  };
  const handleSecondDescendantChange = (event) => {
    const valueEls = getValueEls();
    valueEls.second.textContent = event.target.selected[0] || "";
  };
  return html`
        <sp-action-menu @change=${handleRootChange} @sp-opened=${clearValues}>
            <sp-icon-show-menu slot="icon"></sp-icon-show-menu>
            <sp-menu-group
                @change=${() => console.log("group change")}
                role="none"
            >
                <span slot="header">New York</span>
                <sp-menu-item>Bronx</sp-menu-item>
                <sp-menu-item id="submenu-item-1">
                    Brooklyn
                    <sp-menu
                        slot="submenu"
                        @change=${handleFirstDescendantChange}
                    >
                        <sp-menu-item id="submenu-item-2">
                            Ft. Greene
                            <sp-menu
                                slot="submenu"
                                @change=${handleSecondDescendantChange}
                            >
                                <sp-menu-item>S. Oxford St</sp-menu-item>
                                <sp-menu-item>S. Portland Ave</sp-menu-item>
                                <sp-menu-item>S. Elliot Pl</sp-menu-item>
                            </sp-menu>
                        </sp-menu-item>
                        <sp-menu-item disabled>Park Slope</sp-menu-item>
                        <sp-menu-item>Williamsburg</sp-menu-item>
                    </sp-menu>
                </sp-menu-item>
                <sp-menu-item>
                    Manhattan
                    <sp-menu
                        slot="submenu"
                        @change=${handleFirstDescendantChange}
                    >
                        <sp-menu-item disabled>SoHo</sp-menu-item>
                        <sp-menu-item>
                            Union Square
                            <sp-menu
                                slot="submenu"
                                @change=${handleSecondDescendantChange}
                            >
                                <sp-menu-item>14th St</sp-menu-item>
                                <sp-menu-item>Broadway</sp-menu-item>
                                <sp-menu-item>Park Ave</sp-menu-item>
                            </sp-menu>
                        </sp-menu-item>
                        <sp-menu-item>Upper East Side</sp-menu-item>
                    </sp-menu>
                </sp-menu-item>
            </sp-menu-group>
        </sp-action-menu>
        <div>
            Root value:
            <span id="root-value"></span>
            <br />
            First descendant value:
            <span id="first-value"></span>
            <br />
            Second descendant value:
            <span id="second-value"></span>
            <br />
        </div>
    `;
};
submenu.decorators = [submenuDecorator];
export const contextMenu = () => {
  const contextMenuTemplate = () => html`
        <sp-popover
            style="max-width: 33vw;"
            @click=${(event) => {
    var _a;
    return (_a = event.target) == null ? void 0 : _a.dispatchEvent(
      new Event("close", { bubbles: true })
    );
  }}
        >
            <sp-menu @change=${handleRootChange}>
                <sp-menu-group>
                    <span slot="header">Options</span>
                    <sp-menu-item>
                        Copy
                        <span slot="value">⌘​S</span>
                    </sp-menu-item>
                    <sp-menu-item>
                        Paste
                        <span slot="value">⌘​P</span>
                    </sp-menu-item>
                    <sp-menu-item>
                        Cut
                        <span slot="value">⌘​X</span>
                    </sp-menu-item>
                    <sp-menu-divider></sp-menu-divider>
                    <sp-menu-item>
                        Select layer
                        <sp-menu
                            slot="submenu"
                            selects="single"
                            @change=${handleFirstDescendantChange}
                        >
                            <sp-menu-item selected>Ellipse 1</sp-menu-item>
                            <sp-menu-item>Rectangle</sp-menu-item>
                        </sp-menu>
                    </sp-menu-item>
                    <sp-menu-item>
                        Group
                        <span slot="value">⌘​G</span>
                    </sp-menu-item>
                    <sp-menu-item>
                        Unlock
                        <span slot="value">⌘​L</span>
                    </sp-menu-item>
                    <sp-menu-divider></sp-menu-divider>
                    <sp-menu-item>
                        Bring to front
                        <span slot="value">⇧​⌘​​]</span>
                    </sp-menu-item>
                    <sp-menu-item>
                        Bring forward
                        <span slot="value">⌘​​]</span>
                    </sp-menu-item>
                    <sp-menu-item>
                        Send backward
                        <span slot="value">⌘​​[</span>
                    </sp-menu-item>
                    <sp-menu-item>
                        Send to back
                        <span slot="value">⇧​⌘​​[</span>
                    </sp-menu-item>
                    <sp-menu-divider></sp-menu-divider>
                    <sp-menu-item>
                        Delete
                        <span slot="value">DEL</span>
                    </sp-menu-item>
                </sp-menu-group>
            </sp-menu>
        </sp-popover>
    `;
  const pointerenter = async (event) => {
    event.preventDefault();
    const trigger = event.target;
    const virtualTrigger = new VirtualTrigger(event.clientX, event.clientY);
    const fragment = document.createDocumentFragment();
    render(contextMenuTemplate(), fragment);
    const popover = fragment.querySelector("sp-popover");
    clearValues();
    openOverlay(trigger, "modal", popover, {
      placement: "right-start",
      receivesFocus: "auto",
      virtualTrigger
    });
  };
  const getValueEls = () => {
    return {
      root: document.querySelector("#root-value"),
      first: document.querySelector("#first-value")
    };
  };
  const clearValues = () => {
    const valueEls = getValueEls();
    valueEls.root.textContent = "";
    valueEls.first.textContent = "";
  };
  const handleRootChange = (event) => {
    const valueEls = getValueEls();
    valueEls.root.textContent = event.target.value;
  };
  const handleFirstDescendantChange = (event) => {
    const valueEls = getValueEls();
    valueEls.first.textContent = event.target.selected[0] || "";
  };
  return html`
        <style>
            .app-root {
                position: absolute;
                inset: 0;
            }
            active-overlay::part(theme) {
                --swc-menu-width: 200px;
            }
        </style>
        <div class="app-root" @contextmenu=${pointerenter}>
            <div>
                Root value:
                <span id="root-value"></span>
                <br />
                First descendant value:
                <span id="first-value"></span>
                <br />
            </div>
        </div>
    `;
};
