// @spectrum-web-components/toast/stories/toast.stories.js
"use strict";
import { html } from "@spectrum-web-components/base";
import "@spectrum-web-components/toast/sp-toast.js";
import "@spectrum-web-components/button/sp-button.js";
const toast = ({
  variant = "",
  open = true,
  content = ""
}) => html`
    <sp-toast
        variant=${variant}
        ?open=${open}
    >
        ${content}
        <sp-button slot="action" variant="overBackground" quiet>Undo</sp-button>
    </sp-toast>
`;
export default {
  component: "sp-toast",
  title: "Toast",
  args: {
    content: "This is a toast message.",
    open: true
  },
  argTypes: {
    content: {
      name: "content",
      type: { name: "string", required: false },
      table: {
        type: { summary: "string" },
        defaultValue: { summary: "" }
      },
      control: "text"
    },
    open: {
      name: "open",
      type: { name: "boolean", required: false },
      description: "Whether the toast is open.",
      table: {
        type: { summary: "boolean" },
        defaultValue: { summary: false }
      },
      control: {
        type: "boolean"
      }
    }
  }
};
export const Default = ({
  variant,
  open,
  content
}) => {
  return toast({ variant, open, content });
};
const variantDemo = ({
  variant,
  open,
  content
}) => {
  return toast({ variant, open, content });
};
export const Positive = (args) => variantDemo({ ...args, variant: "positive" });
export const Negative = (args) => variantDemo({ ...args, variant: "negative" });
export const Info = (args) => variantDemo({ ...args, variant: "info" });
