// @swc-react/menu/index.dev.js
"use strict";
import * as React from "react";
import { createComponent } from "@lit-labs/react";
import { Menu as SpMenu } from "@spectrum-web-components/menu";
import { MenuDivider as SpMenuDivider } from "@spectrum-web-components/menu";
import { MenuGroup as SpMenuGroup } from "@spectrum-web-components/menu";
import { MenuItem as SpMenuItem } from "@spectrum-web-components/menu";
import "@spectrum-web-components/menu/sp-menu-divider.js";
import "@spectrum-web-components/menu/sp-menu-group.js";
import "@spectrum-web-components/menu/sp-menu-item.js";
import "@spectrum-web-components/menu/sp-menu.js";
export const Menu = createComponent({
  displayName: "Menu",
  elementClass: SpMenu,
  react: React,
  tagName: "sp-menu",
  events: {
    change: "change",
    close: "close"
  }
});
export const MenuDivider = createComponent({
  displayName: "MenuDivider",
  elementClass: SpMenuDivider,
  react: React,
  tagName: "sp-menu-divider",
  events: {}
});
export const MenuGroup = createComponent({
  displayName: "MenuGroup",
  elementClass: SpMenuGroup,
  react: React,
  tagName: "sp-menu-group",
  events: {
    change: "change",
    close: "close"
  }
});
export const MenuItem = createComponent({
  displayName: "MenuItem",
  elementClass: SpMenuItem,
  react: React,
  tagName: "sp-menu-item",
  events: {
    spMenuItemAdded: "sp-menu-item-added",
    spMenuItemRemoved: "sp-menu-item-removed"
  }
});
